### PR TITLE
[IMPORT-774] Graphql data mode from curl import

### PR DESCRIPTION
**Pull request by aman-v-singh**
- Checking if the Request Body Data matches the graphql format.
- If the request body data is in graphql format then return the graphql body object with mode set to `'graphql'`
- https://github.com/postmanlabs/postman-app-support/issues/9941
- Corresponding changes in [Postman-App](https://github.com/postman-eng/postman-app/pull/5230) repo.

https://github.com/postman-eng/postman-app/assets/121886615/f0892c6c-ef0b-486e-a35b-ddfb28f342ba

### CODE REVIEW

File: src/lib.js [hunk 1]
```diff
@@ -677,6 +676,36 @@ var program,
       }
     },
 
+    /**
+     *
```
Comment (VShingala) [inline src/lib.js:680]: Add what brief infor about what function does here.

For example: https://github.com/postmanlabs/curl-to-postman/blob/b7971d824ffbc6036ea102221a318ce6f3d9aac5/src/lib.js#L494

File: src/lib.js [hunk 2]
```diff
@@ -748,10 +777,19 @@ var program,
...
             bodyArr.push(this.trimQuotesFromString(dataAsciiString));
 
-            request.body.mode = 'raw';
-            request.body.raw = _.join(_.reject(bodyArr, (ele) => {
-              return !ele;
-            }), '&');
+            const rawDataString = _.join(_.reject(bodyArr, (ele) => {
+                return !ele;
+              }), '&'),
+              isGraphqlBody = this.checkIfGraphql(rawDataString);
```
Comment (VShingala) [inline src/lib.js:783]: `isGraphqlBody` variable naming suggests the value to be of type `boolean`. Which is not the case here as it's an `object`. Can we come up with a better name?

File: src/lib.js [hunk 3]
```diff
@@ -677,6 +676,36 @@ var program,
       }
     },
 
+    /**
+     *
+     * @param {string} dataString - Input data string to check if it is a graphql query
+     * @returns {Object} - { result: true, graphql: {Object} } if dataString is a graphql query else { result: false }
+    */
+    checkIfGraphql: function (dataString) {
```
Comment (VShingala) [inline src/lib.js:684]: Function naming suggests value to be of type bolean. i.e. Answer to question like `checkIfGraphql` can be true or false only.

Can we rename this better to suggest what it actually does?

**Review by VShingala — CHANGES_REQUESTED** [review]
@aman-v-singh Let's also add unit tests for the changes we have added here.

File: src/lib.js (hunk 2 above)
Comment (aman-v-singh) [reply src/lib.js]: Changed `isGraphqlBody` to `graphqlRequestData`.

File: src/lib.js (hunk 3 above)
Comment (aman-v-singh) [reply src/lib.js]: Changed `checkIfGraphql` to `identifyGraphqlRequest`.

File: src/lib.js [hunk 4]
```diff
@@ -677,6 +676,40 @@ var program,
...
 
+    /**
+     * Identifies whether the input data string is a graphql query or not
+     *
+     * @param {string} dataString - Input data string to check if it is a graphql query
+     * @returns {Object} - { result: true, graphql: {Object} } if dataString is a graphql query else { result: false }
+    */
+    identifyGraphqlRequest: function (dataString) {
+      try {
+        const rawDataObj = _.attempt(JSON.parse, dataString.replace(/\r\n/g, ''));
```
Comment (VShingala) [inline src/lib.js:687]: @aman-v-singh Why do we need to replace `\r\n`? JSON.parse should take care of such characters.

File: src/lib.js [hunk 5]
```diff
@@ -677,6 +676,40 @@ var program,
...
+        if (rawDataObj && !_.isError(rawDataObj)) {
+          if (!_.has(rawDataObj, 'variables')) {
+            rawDataObj.variables = {};
+          }
+          if (_.keys(rawDataObj).length === 2 && _.has(rawDataObj, 'query') && _.has(rawDataObj, 'variables')) {
```
Comment (VShingala) [inline src/lib.js:692]: @aman-v-singh When we discussed this, I remember there being 3rd optional key as well. Are we not supporting it?

File: src/lib.js [hunk 6]
```diff
@@ -677,6 +676,40 @@ var program,
...
+            if (typeof rawDataObj.query === 'string' && typeof rawDataObj.variables === 'object') {
+              return {
+                result: true,
+                graphql: {
+                  query: rawDataObj.query,
+                  // eslint-disable-next-line max-len
```
Comment (VShingala) [inline src/lib.js:698]: Rather than adding custom rule here, we should fix the actual error. i.e. You can store `JSON.stringify(rawDataObj.variables, null, 2)` as variable as it's used 2 times anyways and use that variable for shorter line.

File: src/lib.js (hunk 4 above)
Comment (aman-v-singh) [reply src/lib.js:687]: @VShingala 2 Reasons:-
- I was facing this error. It seems that [I need to escape the `\`](https://stackoverflow.com/questions/42068/how-do-i-handle-newlines-in-json). 
image

- Postman App expects the query to be without any such characters.

File: src/lib.js (hunk 5 above)
Comment (aman-v-singh) [reply src/lib.js:692]: @VShingala The 3rd optional key was `operationName`, we don't have any place to show `operationName` in the Postman App `graphql` editor, but I will add a check for that as well. `operationName` is only required if multiple operations are present in the query. https://graphql.org/learn/serving-over-http/#post-request

Comment (VShingala) [reply src/lib.js]: @aman-v-singh That's fine if we don't support it in postman app, but if we just ignore it here than even valid graphql bodies will be ignored and detected as raw body.

File: src/lib.js (hunk 4 above)
Comment (VShingala) [reply src/lib.js]: @aman-v-singh Hmm, In that case if we remove `/n` newline chars than wouldn't this also remove the correct newline from query as well? Which would be needed for better indentation?

**Review by VShingala — COMMENTED** [review]
Added few comments, also cURL snippet mentioned here is not getting imported correctly with track that you shared in postman app. i.e. body is selected as non in this case.

https://github.com/postmanlabs/postman-app-support/issues/9941

**Comment by aman-v-singh** [comment]
> Added few comments, also cURL snippet mentioned here is not getting imported correctly with track that you shared in postman app. i.e. body is selected as non in this case.
> 
> [postmanlabs/postman-app-support#9941](https://github.com/postmanlabs/postman-app-support/issues/9941)

@VShingala same error is happening in the default track.

File: src/lib.js (hunk 5 above)
Comment (aman-v-singh) [reply src/lib.js]: @VShingala I have added a check for `operationName`.

File: src/lib.js (hunk 4 above)
Comment (aman-v-singh) [reply src/lib.js]: @VShingala The Postman app's graphql editor handles the indentation well based on the space.
eg:- 
This is the result obtained from curl to postman.
Screenshot 2023-06-26 at 1 42 49 PM
On importing the same curl code snippet we get the correct indentation.

Screenshot 2023-06-26 at 1 45 48 PM

File: src/lib.js [hunk 7]
```diff
@@ -677,6 +676,65 @@ var program,
...
+    /**
+     * Identifies whether the input data string is a graphql query or not
+     *
+     * @param {string} dataString - Input data string to check if it is a graphql query
+     * @param {string} contentType - Content type header value
+     * @returns {Object} - { result: true, graphql: {Object} } if dataString is a graphql query else { result: false }
+    */
+    identifyGraphqlRequest: function (dataString, contentType) {
+      try {
+        const rawDataObj = _.attempt(JSON.parse, dataString.replace(/[\r\n]/g, ''));
```
Comment (VShingala) [inline src/lib.js:688]: @aman-v-singh This regex will remove all individual `\n` and `\r` from data. If we have problem with only `\r\n` than let's only replace that with `\n`.

File: src/lib.js [hunk 8]
```diff
@@ -677,6 +676,65 @@ var program,
...
+        if (contentType === 'application/json' && rawDataObj && !_.isError(rawDataObj)) {
+          if (!_.has(rawDataObj, 'query') || !_.isString(rawDataObj.query)) {
+            return { result: false };
+          }
+          if (!(/^[^]*\{[^]*\}$/g).test(rawDataObj.query)) {
```
Comment (VShingala) [inline src/lib.js:693]: @aman-v-singh What is this regexp supposed to do? Do we need this? As discussed if we can solve this without regexp it'd be better solution as we'll be checking this almost all of times due to body being most used format  `application/json`.

File: src/lib.js (hunk 7 above)
Comment (aman-v-singh) [reply src/lib.js]: https://stackoverflow.com/questions/42068/how-do-i-handle-newlines-in-json?noredirect=1&lq=1

Comment (aman-v-singh) [reply src/lib.js]: @VShingala I have added a function to handle escape characters.

File: src/lib.js (hunk 8 above)
Comment (aman-v-singh) [reply src/lib.js:712]: @VShingala without regex it would be difficult to distinguish between a normal string and a `graphql` query.
Data like `--data '{"query": "test"}' ` will also be opening in the `graphql` tab.

Comment (VShingala) [reply src/lib.js:712]: @aman-v-singh Which is completely fine as we discussed. i.e. We can have this edge case handled in future if we come across it but it is very uncommon to have this scenario. Main objective is to not slow down performance for most of imports to handle few edge case.

File: src/lib.js [hunk 9]
```diff
@@ -677,6 +676,77 @@ var program,
...
+          if (_.has(rawDataObj, 'operationName')) {
+            if (!_.isString(rawDataObj.operationName)) {
+              return { result: false };
+            }
+          }
+          else {
+            rawDataObj.operationName = '';
+          }
+          if (_.keys(rawDataObj).length === 3) {
+            if (_.intersection(_.keys(rawDataObj), ['query', 'variables', 'operationName']).length === 3) {
```
Comment (VShingala) [inline src/lib.js:729]: this condition can be removed as we already know if there were other keys, `_.keys(rawDataObj).length` would be > 3. So above condition already handles this.

File: src/lib.js [hunk 10]
```diff
@@ -677,6 +676,77 @@ var program,
...
+              const graphqlVariables = JSON.stringify(rawDataObj.variables, null, 2);
+              return {
+                result: true,
+                graphql: {
+                  query: rawDataObj.query,
+                  operationName: rawDataObj.operationName,
```
Comment (VShingala) [inline src/lib.js:735]: @aman-v-singh Does postman collection support `operationName` field? If it doesn't we can avoid adding it here.

Comment (aman-v-singh) [reply src/lib.js:735]: @VShingala Maybe it will be useful for when the module is used independently.

Comment (VShingala) [reply src/lib.js]: @aman-v-singh Hmm, I think it's fine as we anyways only select particular field from this result in App so it shouldn't affect v1 format.

**Review by VShingala — APPROVED** [review]
